src/vsphere_cpi/lib/cloud/vsphere/datastore_picker.rb

```diff
@@ -50,25 +50,55 @@ def best_disk_placement(disks)
 
           datastores[ds_name][:disks].push(disk)
           datastores[ds_name][:free_space] -= additional_required_space
-          placement[:migration_size] += additional_required_space if existing_ds_name
-
-          found_placement = true
-          break
+          if existing_ds_name && !migration_size_accounted_for
+            placement[:migration_size] += additional_required_space
+            migration_size_accounted_for = true
+          end
         end
 
-        raise Bosh::Clouds::CloudError, pretty_print_placement_error([disk]) unless found_placement
       end
 
+      placement = filter_all_placement_without_disks(placement)
+
+      raise Bosh::Clouds::CloudError, pretty_print_placement_error(disks) if placement[:datastores].empty?
+
       add_balance_score(placement)
+
+    end
+
+    def filter_all_placement_without_disks(placement)
+      datastore_placements = placement[:datastores]
+
+      iterate_placement = datastore_placements.clone
```

Comment:
you can iterate over same hash as well as update it at the same time, so you don't need to clone